src/v/storage/disk_log_appender.cc

```diff
@@ -106,7 +106,7 @@ disk_log_appender::operator()(model::record_batch& batch) {
 }
 
 ss::future<ss::stop_iteration>
-disk_log_appender::append_batch_to_segment(const model::record_batch& batch) {
```

Comment:
maybe if `model::record_batch` interface would return const reference to decompressed record batch (it can be stored as a `std::optional` it would not be necessary to change this method signature. 
`model::record_batch` interface may look like:
```c++
class record_batch {
public:

bool has_decompressed() const {
  return decompressed.has_value();
}

const model::record_batch& decompressed() const {
  return *decompressed;
}

model::record_batch release_decompressed() && {
   return std::move(*decompressed);
}


private:
std::optional<model::record_batch> decompressed;
}

Reply:
As I understand if I make it an option then I won't be able to share it [here](https://github.com/vectorizedio/redpanda/pull/2435/commits/0d24d7c226cb5356afc48cd9f625836a1ae92896#diff-b54120497a6a46905cf09c97f78217463ad1cf23dfc322cec34b5c1b7b92cd19R611) and would need to copy it and thus increase memory pressure.

Reply:
can we do it like: 
```
    record_batch share() {
        return record_batch(
          _header,
          _records.share(0, _records.size_bytes()),
          _compressed,
          _decompressed.has_value() ? std::optional<model::record_batch>(_decompressed->share()) : std::nullopt);
    }
```

Reply:
good point, replacing `std::optional` with `std::unqiue_ptr<model::record_batch>` will help